src/als/iio.rs

```diff
@@ -26,111 +29,123 @@ pub struct Als {
 }
 
 impl Als {
-    pub fn new(base_path: &str, thresholds: HashMap<u64, String>) -> Result<Self, Box<dyn Error>> {
-        Path::new(base_path)
-            .read_dir()
-            .ok()
-            .and_then(|dir| {
-                dir.filter_map(|e| e.ok())
-                    .find(|e| {
-                        ["als", "acpi-als", "apds9960"].contains(
-                            &fs::read_to_string(e.path().join("name"))
-                                .unwrap_or_default()
-                                .trim(),
-                        )
-                    })
-                    .and_then(|e| {
-                        // TODO should probably start from the `parse_illuminance_input` in the next major version
-                        parse_illuminance_raw(e.path())
-                            .or_else(|_| parse_illuminance_input(e.path()))
-                            .or_else(|_| parse_intensity_raw(e.path()))
-                            .or_else(|_| parse_intensity_rgb(e.path()))
-                            .ok()
-                    })
-            })
-            .map(|sensor| Self { sensor, thresholds })
-            .ok_or_else(|| "No iio device found".into())
+    pub async fn new(base_path: &str, thresholds: HashMap<u64, String>) -> Result<Self, ErrorBox> {
+        let mut dir_stream = smol::fs::read_dir(base_path)
+            .await
+            .map_err(|e| ErrorBox::from(format!("Can't enumerate iio devices: {e}")))?;
+
+        while let Some(Ok(e)) = dir_stream.next().await {
```

Comment:
This will make the loop abort if we fail to read a subdirectory wheras on a prior revision, the loop would continue.

Reply:
Great catch! I couldn't resist trying to play with more of the monadic helpers, primarily because I find it beautiful how it manages to express the desirable trajectory (read the dir -> read the sensor names -> skip non-als -> try different parsers -> etc) without having to write explicit loops and conditions. But let me know if you have any feedback on the new approach, I find it very insightful!